aiter/ops/triton/rmsnorm.py

```diff
@@ -834,7 +1036,135 @@ def rmsnorm2d_fwd_with_add(
         NUM_PRGMS,
     )
 
-    return out, residual_out
+    return rsigma
+
+
+def _rmsnorm_backward(dz, x, gamma, rsigma):
+    dz_ = dz.contiguous()
+    x_ = x.contiguous()
+    gamma_ = gamma.contiguous()
+    rsigma_ = rsigma.contiguous()
+
+    dx = torch.empty_like(x_)
+    dgamma = torch.empty_like(gamma_)
+
+    M, N = x_.shape
+    blk_size = block_size(x_)
+    USE_BLOCKED = use_blocked(x_)
+    NUM_PRGMS = num_programs(x_)
+    need_reduction = N > 1
+
+    dg_tmp = (
+        torch.empty(
+            dg_tmp_rows(x_), N, device="cuda", dtype=torch.float32, requires_grad=False
+        )
+        if need_reduction
+        else None
+    )
+
+    grid_bwd = lambda meta: (NUM_PRGMS,)  # noqa: E731
+    _rmsnorm_bwd_triton[grid_bwd](
+        dz_,
+        x_,
+        gamma_,
+        rsigma_,
+        dx,
+        dg_tmp if need_reduction else dgamma,
+        x_.stride(0),
+        dz_.stride(0),
+        M,
+        N,
+        blk_size,
+        USE_BLOCKED,
+        NUM_PRGMS,
+        num_warps=8,
+    )
+
+    if need_reduction:
+        grid_reduce = lambda meta: [triton.cdiv(N, meta["BLOCK_SIZE_N"])]  # noqa: E731
+        _rmsnorm_bwd_dg_reduce_triton[grid_reduce](
+            dg_tmp,
+            dgamma,
+            dg_tmp.stride(0),
+            dg_tmp.shape[0],
+            dg_tmp.shape[1],
+            BLOCK_SIZE_M=128,
+            BLOCK_SIZE_N=64,
+        )
+
+    return dx, dgamma
+
+
+class RMSNorm(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, x, weight, epsilon, is_grad_enabled):
+
+        is_grad = is_grad_enabled and any(
+            tensor.requires_grad for tensor in [x, weight]
+        )
+
+        y, rsigma = _rmsnorm_forward(x, weight, epsilon)
+
+        if is_grad:
+            ctx.save_for_backward(x, weight, rsigma)
+
+        return y
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        x, weight, rsigma = ctx.saved_tensors
+
+        dx, dg = _rmsnorm_backward(grad_output, x, weight, rsigma)
+
+        return dx, dg, None, None
+
+
+class RMSNorm2dFwdWithAdd(torch.autograd.Function):
```

Comment:
Please add _ ie make it _RMSNorm2dFwdWithAdd. Basically make any function and class name start with underscore if it's not meant to be called by the user.